components/resource_metering/src/reporter.rs

```diff
@@ -119,54 +125,84 @@ impl Runnable for ResourceMeteringReporter {
                 self.config = new_config;
             }
             Task::CpuRecords(records) => {
-                let timestamp_secs = records.begin_unix_time_secs;
+                let max_resource_groups = self.config.max_resource_groups;
+                let tmp_group_map = &mut self.tmp_group_map;
+                let tmp_top_vec = &mut self.tmp_top_vec;
+                tmp_group_map.clear();
+                tmp_top_vec.clear();
 
+                // Sum the CPU time by tag
                 for (tag, ms) in &records.records {
                     let tag = &tag.infos.extra_attachment;
                     if tag.is_empty() {
                         continue;
                     }
 
-                    let ms = *ms as u32;
-                    match self.records.get_mut(tag) {
-                        Some((ts, cpu_time, total)) => {
-                            if *ts.last().unwrap() == timestamp_secs {
-                                *cpu_time.last_mut().unwrap() += ms;
-                            } else {
-                                ts.push(timestamp_secs);
-                                cpu_time.push(ms);
-                            }
-                            *total += ms;
-                        }
-                        None => {
-                            self.records
-                                .insert(tag.clone(), (vec![timestamp_secs], vec![ms], ms));
-                        }
-                    }
+                    *tmp_group_map.entry(tag.clone()).or_insert(0) += *ms as u32;
```

Comment:
How about using an Arc to wrap the resource group to reduce memory allocation?

Reply:
It sounds to be complicated. 

It's a good option to use references, but we may either:
1. not to reuse `tmp_group_map` and `tmp_group_vec` or
2. to introduce some `unsafe`s

I prefer `1`: two allocations for eliminating others